### PR TITLE
build(macos)!: add homebrew formula and drop dmg

### LINKED ISSUE #186: Improve MacOS packaging
### Describe the Bug

The dmg packages are not use-able currently and therefore will not be included in release assets until this is resolved.

Known issues:

- missing libraries (will update this later with more information)
  - `libopus.0.dylib` (see below)
- bundle doesn't point to anything, `CPACK_BUNDLE_STARTUP_COMMAND` is not set

```txt
dyld[1028]: Library not loaded: /usr/local/opt/opus/lib/libopus.0.dylib
  Referenced from: /Users/reenignearcher/Downloads/sunshine-macos/Sunshine/usr/local/Sunshine.app/Contents/MacOS/sunshine-0.13.0
  Reason: tried: '/usr/local/opt/opus/lib/libopus.0.dylib' (no such file), '/usr/local/lib/libopus.0.dylib' (no such file), '/usr/lib/libopus.0.dylib' (no such file)
zsh: abort      ./sunshine
```

### Expected Behavior

_No response_

### Additional Context

We could set `CPACK_BUNDLE_STARTUP_COMMAND` similar to this project: https://github.com/MegaGlest/megaglest-source/blob/1e4441e0996c2045f29998a3fd9b2b7a6c2ad441/mk/macos/CMakeLists.txt#L43

The command points to this file: https://github.com/MegaGlest/megaglest-source/blob/develop/mk/macos/bundle_resources/MegaGlest.sh

### Sunshine Host Operating System and Version

MacOS

### Architecture

-

### Sunshine Version

0.14.0

### GPU Type

-

### GPU Model

-

### GPU Driver/Mesa Version

-

### Capture Method (Linux Only)

-

### LINKED ISSUE #342: Any interest in a Homebrew formula for sunshine
### Describe the Bug

Might you be interested in a contributed Homebrew formula for Mac that builds sunshine? 

I haven't had the chance to additionally add a Homebrew service config to the formula and don't think I'll have the chance to circle back to this near term, so wanted to offer up in case useful in the meanwhile.

Formula attached.
[sunshine.rb.txt](https://github.com/LizardByte/Sunshine/files/9418439/sunshine.rb.txt)

### Expected Behavior

_No response_

### Additional Context

_No response_

### Host Operating System

macOS

### Operating System Version

12.5.1

### Architecture

x86_64

### Sunshine Version

0.14.1

### GPU Type

AMD

### GPU Model

Radeon Pro Vega 64 16 GB

### GPU Driver/Mesa Version

01.01.114

### Capture Method (Linux Only)

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by ReenigneArcher**
## Description

This PR will add a homebrew forumla for Sunshine.

Todo:

- [x] Adjust CI and CMake to configure the brew spec file, and build sunshine using the spec (similar to the Portfile)
- [x] Add a LizardByte homebrew repo, similar to https://github.com/LizardByte/jellyfin-plugin-repo
  - Won't hold the action, as brew requires the forumlas to be in the main branch
  - Done, see https://github.com/LizardByte/homebrew (action will take care of creating repo structure)
- [x] Create an action to use for publishing the final homebrew spec to the repo created
- [x] Update docs, specifically macOS installation instructions

References:
- https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap
- https://github.com/realqhc/homebrew

### Issues Fixed or Closed

- Fixes #186 
- Resolves #342 
- Resolves #391 

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update (updates to dependencies)
- [x] Documentation update (changes to documentation)
- [x] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

## Branch Updates
LizardByte requires that branches be up-to-date before merging. This means that after any PR is merged, this branch
must be updated before it can be merged. You must also
[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I want maintainers to keep my branch updated

### CODE REVIEW

File: packaging/macos/sunshine.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  depends_on "node"
+  depends_on "openssl"
+  depends_on "opus"
+
+  def install
+    # initialize submodules
+    system "git", "submodule", "update", "--remote", "--init", "--recursive", "--depth", "1"
+
+    # todo: fix vite build
+    system "sed", "-i", "''", "s/SUNSHINE_SOURCE_ASSETS_DIR=${SUNSHINE_SOURCE_ASSETS_DIR} SUNSHINE_ASSETS_DIR=${CMAKE_BINARY_DIR}//", "cmake/targets/common.cmake"
```
Comment (ReenigneArcher) [inline packaging/macos/sunshine.rb:26]: Todo

**Comment by ReenigneArcher** [comment]
I think we should also drop the standard macOS build as part of this. It will save 3 runners, and the artifacts produced by the build are not actually usable due to missing deps.

The homebrew install will be very easy compared to the macports build as well... essentially it will be a single terminal command to install sunshine through homebrew.

File: cmake/targets/common.cmake [hunk 2]
```diff
@@ -37,8 +37,19 @@ endif()
...
+    set(NPM_ASSETS_DIR ${CMAKE_BINARY_DIR})
+    set(NPM_BUILD_HOMEBREW "")
+endif()
+
 #WebUI build
 add_custom_target(web-ui ALL
         WORKING_DIRECTORY "${CMAKE_SOURCE_DIR}"
         COMMENT "Installing NPM Dependencies and Building the Web UI"
-        COMMAND bash -c \"npm install && SUNSHINE_SOURCE_ASSETS_DIR=${SUNSHINE_SOURCE_ASSETS_DIR} SUNSHINE_ASSETS_DIR=${CMAKE_BINARY_DIR} npm run build\") # cmake-lint: disable=C0301
+        COMMAND bash -c \"npm install && SUNSHINE_BUILD_HOMEBREW=${NPM_BUILD_HOMEBREW} SUNSHINE_SOURCE_ASSETS_DIR=${NPM_SOURCE_ASSETS_DIR} SUNSHINE_ASSETS_DIR=${NPM_ASSETS_DIR} npm run build\") # cmake-lint: disable=C0301
```
Comment (ReenigneArcher) [inline cmake/targets/common.cmake:55]: vite build fails if we provide a full path to the directories

File: packaging/macos/sunshine.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
+require "language/node"
+
+class @PROJECT_NAME@ < Formula
+  desc "@PROJECT_DESCRIPTION@"
+  homepage "@PROJECT_HOMEPAGE_URL@"
+  url "@GITHUB_CLONE_URL@",
+    tag: "@GITHUB_BRANCH@"
+  version "@PROJECT_VERSION@"
+  license all_of: ["GPL-3.0-only"]
+  head "@GITHUB_CLONE_URL@"
```
Comment (ReenigneArcher) [inline packaging/macos/sunshine.rb:10]: The file is configured just like the Portfile, flatpak, etc.

File: packaging/macos/sunshine.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+  depends_on "boost" => :build
+  depends_on "cmake" => :build
+  depends_on "pkg-config" => :build
+  depends_on "curl"
+  depends_on "miniupnpc"
+  depends_on "node"
+  depends_on "openssl"
+  depends_on "opus"
```
Comment (ReenigneArcher) [inline packaging/macos/sunshine.rb:19]: Should we add avahi?

File: packaging/macos/sunshine.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  end
+
+  service do
+    run [opt_bin/"sunshine", "~/.config/sunshine/sunshine.conf"]
+  end
+
+  test do
+    # test that the binary runs at all
+    output = shell_output("#{bin}/sunshine --version").strip
+    puts output
```
Comment (ReenigneArcher) [inline packaging/macos/sunshine.rb:44]: without this the output is suppressed

**Comment by ReenigneArcher** [comment]
@realhq would you mind reviewing this?

homebrew repo will be here: https://github.com/LizardByte/homebrew

The action that publishes the tap has been tested here: https://github.com/LizardByte/homebrew-release-action/tree/tests

New docs for homebrew are here: https://lizardbyte--2222.org.readthedocs.build/projects/sunshine/en/2222/about/setup.html#install

And the generated spec file can be seen in the CI here: https://github.com/LizardByte/Sunshine/actions/runs/8195989432/job/22415345732?pr=2222#step:4:61

File: packaging/macos/sunshine.rb (hunk 4 above)
Comment (XiongjieDai) [reply packaging/macos/sunshine.rb:19]: when I try `brew install avahi`
avahi: Linux is required for this software.

**Comment by ReenigneArcher** [comment]
The core parts of this PR will remain the same, but...

I think I'm going to modify the action to do more of the heavy lifting.

1. Provide it a specific formula as an input
2. Have the action run the install
3. Have the action run the tests
4. Audit the formula (https://docs.brew.sh/Formula-Cookbook#audit-the-formula)

And for the future:

5. Fork https://github.com/Homebrew/homebrew-core if it doesn't exist in the org already
6. Checkout the fork
7. Create a branch for the formula, and commit the changes
8. Create/update a PR to upstream

**Review by cgutman — APPROVED** [review]
Seems fine to me, though I'm by no means an expert in Homebrew packaging.

File: CMakeLists.txt [hunk 6]
```diff
@@ -4,7 +4,7 @@ cmake_minimum_required(VERSION 3.18)
 
 # todo - set version to 0.0.0 once confident in automated versioning
 project(Sunshine VERSION 0.22.0
-        DESCRIPTION "Sunshine is a self-hosted game stream host for Moonlight."
+        DESCRIPTION "Self-hosted game stream host for Moonlight"
```
Comment (ReenigneArcher) [inline CMakeLists.txt:7]: This was modified to appease the homebrew audit step. The same complaints were presented by the Macports team previously, and I think the Flatpak team.

**Comment by tristan-k** [comment]
Why isn't the thomebrew formula up to date with the latest [sunshine.rb release ](https://github.com/LizardByte/Sunshine/releases/download/v2024.719.161459/sunshine.rb)? There is also the [homebrew head flag](https://github.com/Homebrew/brew/blob/master/docs/Formula-Cookbook.md#unstable-versions-head) in case that it isn't updated because it is a beta releease.

**Comment by ReenigneArcher** [comment]
head only works if the build process hasn't changed

**Comment by tristan-k** [comment]
What a bummer. Is there another way to use the `sunshine.rb` with homebrew or some documentation?

**Comment by ReenigneArcher** [comment]
Latest stable version should work with brew

**Pull request merged**